src/gtc/passes/oir_optimizations/inlining.py

```diff
@@ -0,0 +1,112 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import copy as cp
+from typing import Any, Dict, Optional
+
+from eve import NodeTranslator
+from gtc import oir
+
+
+class MaskInlining(NodeTranslator):
+    def visit_FieldAccess(
+        self,
+        node: oir.FieldAccess,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.Expr:
+        if node.name in masks_to_inline:
+            return cp.deepcopy(masks_to_inline[node.name])
+        return self.generic_visit(node, masks_to_inline=masks_to_inline, **kwargs)
+
+    def visit_AssignStmt(
+        self,
+        node: oir.AssignStmt,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.AssignStmt:
+        target_name = node.left.name
+        if target_name in masks_to_inline:
+            masks_to_inline[target_name] = node.right
+            return None
+        return self.generic_visit(node, masks_to_inline=masks_to_inline, **kwargs)
+
+    def visit_MaskStmt(
+        self,
+        node: oir.MaskStmt,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.MaskStmt:
+        return oir.MaskStmt(
+            mask=self.visit(node.mask, masks_to_inline=masks_to_inline, **kwargs),
+            body=self.visit(node.body, masks_to_inline=masks_to_inline, **kwargs),
+        )
+
+    def visit_HorizontalExecution(
+        self,
+        node: oir.HorizontalExecution,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.HorizontalExecution:
+        return oir.HorizontalExecution(
+            body=[
+                stmt
+                for stmt in self.visit(
+                    node.body,
+                    masks_to_inline=masks_to_inline,
+                    **kwargs,
+                )
+                if stmt is not None
+            ],
+            declarations=node.declarations,
+        )
+
+    def visit_VerticalLoop(
+        self,
+        node: oir.VerticalLoop,
+        masks_to_inline: Dict[str, Optional[oir.Expr]],
+        **kwargs: Any,
+    ) -> oir.VerticalLoop:
+        local_masks_to_inline = (
+            node.iter_tree()
+            .if_isinstance(oir.FieldAccess)
+            .getattr("name")
+            .filter(lambda name: name.startswith("mask_"))
```

Comment:
We don't encode semantic in names, it might work today, but could break anytime. (I actually don't know if a user-defined field with name `match_something` could accidentally be matched here. Instead just check fields in mask statements.